File: src/adapters/github/issues-notifier.js
"use strict";

const { parseIso } = require("../../engine/timeutil");

const API_ROOT = "https://api.github.com";
const FAILURE_LABEL = "cron-failure";

class GitHubIssuesNotifier {
  constructor(ledger, repo, token) {
    this.ledger = ledger;
    this.repo = repo;
    this.token = token;
  }

  headers() {
    return {
      Authorization: `Bearer ${this.token}`,
      Accept: "application/vnd.github+json",
      "X-GitHub-Api-Version": "2022-11-28",
      "Content-Type": "application/json",
    };
  }

  async notifyFailure(job, result, consecutiveFailures) {
    const body =
      `Job \`${job.id}\` failed.\n\n` +
      `- Scheduled: ${result.scheduled_time}\n` +
      `- Attempts: ${result.attempts}\n` +
      `- Detail: ${result.detail}\n` +
      `- Consecutive failures: ${consecutiveFailures}\n\n` +
      "See the Actions run and dashboard for more.";
    try {
      const existing = this.ledger.openIssueNumber(job.id);
      if (existing) {
        await this.comment(existing, body);
      } else {
        const number = await this.createIssue(job, body);
        if (number !== null) this.ledger.recordOpenIssue(job.id, number, parseIso(result.finished_at));
      }
    } catch (err) {
      // Best-effort notification: never fail the dispatcher after job execution.
    }
  }

  async notifyRecovery(job, result) {
    const existing = this.ledger.openIssueNumber(job.id);
    if (!existing) return;
    try {
      await this.comment(existing, `Recovered: job \`${job.id}\` succeeded at ${result.finished_at}.`);
      await this.close(existing);
      this.ledger.clearIssue(job.id);
    } catch (err) {
      // Best-effort notification.
    }
  }

  async createIssue(job, body) {
    const response = await fetch(`${API_ROOT}/repos/${this.repo}/issues`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify({
        title: `[cron] ${job.id} failing`,
        body,
        labels: [FAILURE_LABEL, `job:${job.id}`],
      }),
      signal: AbortSignal.timeout(15000),
    });
    if (response.status >= 300) return null;
    const data = await response.json();
    return data.number ?? null;
  }

  async comment(issueNumber, body) {
    await fetch(`${API_ROOT}/repos/${this.repo}/issues/${issueNumber}/comments`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify({ body }),
      signal: AbortSignal.timeout(15000),
    });
  }

  async close(issueNumber) {
    await fetch(`${API_ROOT}/repos/${this.repo}/issues/${issueNumber}`, {
      method: "PATCH",
      headers: this.headers(),
      body: JSON.stringify({ state: "closed", state_reason: "completed" }),
      signal: AbortSignal.timeout(15000),
    });
  }
}

module.exports = { GitHubIssuesNotifier };

File: adapters/github/issues_notifier.py
"""GitHub Issues adapter implementing engine.notify.Notifier. One open issue
per failing job (comments accumulate on it, never spammed with new issues),
auto-closed with a recovery comment on the next success. Issue-number
tracking lives in the shared Ledger (state/issues.json) so it survives across
runs regardless of which notifier backend is in use.

Best-effort: a notification failure (network blip, rate limit) never fails
the dispatcher run — job execution and state persistence already happened by
the time this is called.
"""
from __future__ import annotations

import requests

from engine.ledger import Ledger
from engine.models import Job, RunResult
from engine.timeutil import parse_iso

API_ROOT = "https://api.github.com"
FAILURE_LABEL = "cron-failure"


class GitHubIssuesNotifier:
    def __init__(self, ledger: Ledger, repo: str, token: str):
        self.ledger = ledger
        self.repo = repo
        self.session = requests.Session()
        self.session.headers.update(
            {
                "Authorization": f"Bearer {token}",
                "Accept": "application/vnd.github+json",
                "X-GitHub-Api-Version": "2022-11-28",
            }
        )

    def notify_failure(self, job: Job, result: RunResult, consecutive_failures: int) -> None:
        body = (
            f"Job `{job.id}` failed.\n\n"
            f"- Scheduled: {result.scheduled_time}\n"
            f"- Attempts: {result.attempts}\n"
            f"- Detail: {result.detail}\n"
            f"- Consecutive failures: {consecutive_failures}\n\n"
            f"See the Actions run and dashboard for more."
        )
        try:
            existing = self.ledger.open_issue_number(job.id)
            if existing:
                self._comment(existing, body)
            else:
                number = self._create_issue(job, body)
                if number is not None:
                    self.ledger.record_open_issue(job.id, number, parse_iso(result.finished_at))
        except requests.exceptions.RequestException:
            pass

    def notify_recovery(self, job: Job, result: RunResult) -> None:
        existing = self.ledger.open_issue_number(job.id)
        if not existing:
            return
        try:
            self._comment(existing, f"Recovered: job `{job.id}` succeeded at {result.finished_at}.")
            self._close(existing)
            self.ledger.clear_issue(job.id)
        except requests.exceptions.RequestException:
            pass

    def _create_issue(self, job: Job, body: str) -> int | None:
        resp = self.session.post(
            f"{API_ROOT}/repos/{self.repo}/issues",
            json={
                "title": f"[cron] {job.id} failing",
                "body": body,
                "labels": [FAILURE_LABEL, f"job:{job.id}"],
            },
            timeout=15,
        )
        if resp.status_code >= 300:
            return None
        return resp.json().get("number")

    def _comment(self, issue_number: int, body: str) -> None:
        self.session.post(
            f"{API_ROOT}/repos/{self.repo}/issues/{issue_number}/comments",
            json={"body": body},
            timeout=15,
        )

    def _close(self, issue_number: int) -> None:
        self.session.patch(
            f"{API_ROOT}/repos/{self.repo}/issues/{issue_number}",
            json={"state": "closed", "state_reason": "completed"},
            timeout=15,
        )
